### ISSUE TITLE
Jemalloc unconditionally leaks

### ISSUE THREAD

**Issue by sbufalov** [jemalloc/jemalloc#730]
Jemalloc built with default options leaks inside 'malloc_init'. It looks like a generic deficiency in the product and can be reproduced at least on windows, linux, and osx.

**Comment by jasone** [comment]
There is not enough information here to reproduce the issue.  Furthermore, it is likely that what you are calling a leak is actually jemalloc creating metadata structures, which is expected behavior.

**Comment by sbufalov** [comment]
Yeah. I know that jemalloc is allocating metadata structures in the call to 'malloc_init'. What I don't expect is that jemalloc doesn't deallocate metadata structures at exit.

The reproduction case for the reported problem is amazingly simple! Build jemalloc shared library with default settings and create an application that just loads and immediately unloads the library multiple times. The larger is the amount of loads/unloads the larger is the application's memory footprint because at each load/unload jemalloc allocates more and more and frees nothing (see the sample code below).

I can imagine that jemalloc would instantiate the metadata once in a running process. And even if it doesn't release the metadata latter it would be a constant memory leak.

The issue is closed. Does it mean that the problem will not be addressed at all?

```
#include <stdio.h>
#include <dlfcn.h>
#include <sys/time.h>
#include <sys/resource.h>
const char* library_name = "libjemalloc.so";
void load_library() {
    void* handle = dlopen(library_name, RTLD_NOW);
    if (handle)
        dlclose(handle);
}
int main(int argc, char* argv[]) {
    struct rusage ru;
    load_library();
    getrusage(RUSAGE_SELF, &ru);
    printf("MAX RSS usage %ld after first load and unload libarary: %s\n", ru.ru_maxrss, library_name);
    int count = 1000;
    for (int i = 0; i < count; i++)
        load_library();
    getrusage(RUSAGE_SELF, &ru);
    printf("MAX RSS usage %ld after %d load and unload libarary: %s\n", ru.ru_maxrss, count, library_name);
    return 0;
}
```

**Comment by davidtgoldblatt** [comment]
This isn't a usage pattern I find particularly compelling -- is this actually causing a problem for you? Can you go into a little bit more detail about whatever you're actually trying to solve?

**Comment by sbufalov** [comment]
We found jemalloc to be a very good product in terms of speed and features. In the last release we begun to use it as a backend for memory allocations in our HPC components on SMP and NUMA systems.

Of cause, the example above is just an exaggeration. But it describes the problem pretty well.
In real life we have a number of cases when an application needs to load and unload our component several times in a single run of a process. It's not thousands of loads and unloads, but even with a single one the memory leaks.

Jemalloc causes memory leaks when it is statically linked in our shared library as well.

Our QA acceptance criteria doesn't allow to have memory leaks in our product. If there is any option which can help us to get metadata deallocated at the end we are ready to consider it.

We will highly appreciate any hint on this topic from your side! Thank you!

**Comment by davidtgoldblatt** [comment]
A couple more questions:
- Are you able to refactor the way you ship libraries you ship so that jemalloc need only be loaded once, even if your component is loaded and unloaded many times?
- Could you describe a little bit more about the application architecture?

The problem on our end is that the environments we build and test for all sort of assume our lifetime is process-scoped, and backing out that assumption (and testing for it in the future) is going to be pretty tricky.

We might be able to expose a "no, really, I'm not kidding, shut down everything" function that cleans up our metadata and all allocations done through the library as an experimental API (see #567) and let the responsibility for calling that fall to your library -- would that be enough for your use case?

Reopening as a feature request, but I don't want to overstate the likelihood of a fix coming from our end, especially in the short term. The 5.2 release is the earliest this could happen, and even that is if the blockers fall into place around it.

**Comment by sbufalov** [comment]
We are working in a media processing/delivery industry. We are not implementing applications ourselves. We provide highly optimized cross-platform low-level libraries. Jemalloc is believed to help us to efficiently manage memory allocations on NUMA and HBM.

What you suggested as a part of experimental API (#567) sounds like a minimum required functionality that we can try to use to resolve our issue. It should be enough. On top of it, if required, we can think about some small redesign of how jemalloc is handled in our libraries.

How do we know that "no, really, I'm not kidding, shut down everything" got implemented? We want to try it individually ASAP.

Thank you for reopening this request. We plan to continue using jemalloc. So any jemalloc improvements made in the future even far future are improvements in our product as well.

**Comment by davidtgoldblatt** [comment]
I'll try to ping this thread if we implement it (or get ourselves into a state where it's easily implementable). In the long run, I'd like to refactor out as much of our global state as possible just to make testing easier, but as a side effect it should make it much more straightforward to write a teardown function.

**Comment by martin-schink** [comment]
Is it possible to guide our developers to implement such an "stop all ship, de-allocate all" API in a fork of the current master? If you could sketch out where to look and and what structures to de-allocate that gives us at least another option to temporarily work around this issues for our release in three weeks.  Afterwards we can for sure discuss if it makes sense to contribute the code and tests to the project, and what other code quality guidelines must be adhered to - or if it's just a temporary fix that merely serves as a prototype which will be discarded. As it stands right now the team feels uneasy digging into unknown code to attempt a fix in such a short time. If you'd like we could maybe arrange for a small call to keep your investment minimal.

In short: if you help us fix this on our own, we'd like to aid the product development and ensure the use case is properly supported in future releases.

**Comment by interwq** [comment]
In addition to what David mentioned, I believe currently all the auto arena structs are also never destroyed. Currently on dev branch there is arena.destroy interface, but it applies to the extended arenas only. We may want to make the destroy API work for auto arenas as well (unfortunately not in short term).

**Comment by martin-schink** [comment]
Thanks, that helps the team to get started :)

**Comment by YongMan** [comment]
@sbufalov In my use case, I make redis(which use jemalloc for memory management) hot reloadable use dlopen/dlclose, I also meet this problem.
I link jemalloc dynamic library in main thread, and jemalloc will just init once, and will not re-init duiring dlopen .so file.

**Comment by sbufalov** [comment]
I have continued with the approach suggested by @davidtgoldblatt. It works for me!
By the way, I link jemalloc statically.
